Documentation/design-docs/standalone-gc-loading.md

```diff
@@ -0,0 +1,222 @@
+
+## Definitions
+
+Before diving in to the specification, it's useful to precisely define
+some terms that will be used often in this document.
+
+* The **Execution Engine**, or **EE** - The component of the CLR responsible for *executing* programs.
+  This is an intentionally vague definition. The GC does not care how (or even *if*) programs are
+  compiled or executed, so it is up to the EE to invoke the GC whenever an executing
+  program does something that requires the GC's attention. The EE is notable because the implementation
+  of an execution engine varies widely between runtimes; the CoreRT EE is primarily in managed code
+  (C#), while CoreCLR (and the .NET Framework)'s EE is primarily in C++.
+* The **GC**, or **Garbage Collector** - The component of the CLR responsible for allocating managed
+  objects and reclaming unused memory. It is written in C++ and the code is shared by multiple runtimes.
+  (That is, CoreCLR/CoreRT may have different execution engines, but they share the *same* GC code.)
+* The **DAC**, or **Data Access Component** - A subset of the execution engine that is compiled in
+  such a way that it can be run *out of process*, when debugging .NET code using a debugger. The DAC
+  is used by higher-level components such as SOS (Son of Strike, a windbg/lldb debugger extension for
+  debugging managed code) and the DBI (a COM interface). The full details about the DAC are covered in
+  [this document](https://github.com/dotnet/coreclr/blob/master/Documentation/botr/dac-notes.md).
+
+## Rationale and Goals of a Standalone GC
+
+A GC that is "standalone" is one that is able to be built as a dynamic shared library and loaded
+dynamically at startup. This definition is useful because it provides a number of benefits
+to the codebase:
+
+* A standalone GC that can be loaded dynamically at runtime can also be *substituted* easily by
+  loading any GC-containing dynamic shared library specified by a user. This is especially interesting
+  for prototyping and testing GC changes since it would be possible to make changes to the GC
+  without having to re-compile the runtime.
+* A standalone GC that can be *built* as a dynamic shared library imposes a strong requirement that
+  the interfaces that the GC uses to interact with other runtime components be complete and
+  correct. A standalone GC will not link successfully if it refers to symbols defined within
+  the EE. This makes the GC codebase significantly easier to share between different execution
+  engine implementations; as long as the GC implements its side of the interface and the EE
+  implements its side of the interface, we can expect that changes within the GC itself
+  will be trivially portable to other runtime implementations.
+
+  This is a significant engineering pain point today. A great deal of time is spent manually porting
+  changes between the different branches of the GC codebase. When changes are not ported, the GC
```

Comment:
FWIW, you still need to do the work to flow the changes between different products/branches/releases even with standalone GC. It may be easier with enforced interface, but it is still going to take a great deal of time. 

In the RyuJIT case, the maintenance of JIT/EE interface and SuperPMI is non-trivial effort.

Reply:
fair enough, I did not mean to imply that. I'll change the wording.